serving/src/main/java/feast/serving/specs/CachedSpecService.java

```diff
@@ -103,69 +103,99 @@ public FeatureSetSpec getFeatureSetSpec(String featureSetRef) throws ExecutionEx
-                            + "Please specify a more specific Feature Reference (ie specify the project or feature set)",
-                        getFeatureStringWithProjectRef(featureReference)));
+              // apply project override when finding feature set for feature
+              FeatureReference queryFeatureRef = featureReference;
+              if (!project.isEmpty()) {
+                queryFeatureRef = featureReference.toBuilder().setProject(project).build();
               }
+
+              String fsName = mapFeatureToFeatureSetName(queryFeatureRef);
               return Pair.of(fsName, featureReference);
             })
         .collect(groupingBy(Pair::getLeft))
         .forEach(
-            (fsName, featureRefs) -> {
-              try {
-                FeatureSetSpec featureSetSpec = featureSetCache.get(fsName);
-                List<FeatureReference> requestedFeatures =
-                    featureRefs.stream().map(Pair::getRight).collect(Collectors.toList());
-
-                // check that requested features reference point to different features in the
-                // featureset.
-                HashSet<String> featureNames = new HashSet<>();
-                requestedFeatures.forEach(
-                    ref -> {
-                      if (featureNames.contains(ref.getName())) {
-                        throw new SpecRetrievalException(
-                            "Multiple Feature References referencing the same feature in a featureset is not allowed.");
-                      }
-                      featureNames.add(ref.getName());
-                    });
-
-                FeatureSetRequest featureSetRequest =
-                    FeatureSetRequest.newBuilder()
-                        .setSpec(featureSetSpec)
-                        .addAllFeatureReferences(requestedFeatures)
-                        .build();
-                featureSetRequests.add(featureSetRequest);
-              } catch (ExecutionException e) {
-                throw new SpecRetrievalException(
-                    String.format("Unable to find featureSet with name: %s", fsName), e);
-              }
+            (fsName, fsNamesAndfeatureRefs) -> {
+              List<FeatureReference> featureRefs =
+                  fsNamesAndfeatureRefs.stream().map(Pair::getRight).collect(Collectors.toList());
+              featureSetRequests.add(buildFeatureSetRequest(fsName, featureRefs));
             });
     return featureSetRequests;
   }
 
+  /** Build a Feature Set request from the Feature Set specified by name and Feature References */
+  private FeatureSetRequest buildFeatureSetRequest(
+      String featureSetName, List<FeatureReference> featureReferences) {
+    // get feature set for name
+    FeatureSetSpec featureSetSpec;
+    try {
+      featureSetSpec = featureSetCache.get(featureSetName);
+    } catch (ExecutionException e) {
```

Comment:
One of the major pain point that I have seen is when users are requesting features that dont exist because a store is not subscribed to it. It maybe out of scope for this PR, but is there a way that we can address that problem?

Basically they should get an exception that a feature set exists but is not being populated, instead of just "this feature or feature set doesnt exist".

Reply:
I think one way we can resolve this populate the feature set cache from core wholesale and apply the subscription checking only when request is made. This way we be able to determine if the request is rejected due missing feature set or because subscription was not applied properly.